Merge dng vis

anderspitman opened the pull request:
This is a test. I merged the git repos per [this SO answer](http://stackoverflow.com/a/10548919/943814). This retains all the commits I made on the original dng-vis repo. If we'd rather just copy the files as they are now and start with a fresh history that's fine.

reedacartwright committed:
Merge branch 'release/1.1.1'

RELEASE NOTES

* FEATURE: When building with cmake, users have the option to download and install dependencies.
Just run `cmake -DBUILD_EXTERNAL_PROJECTS=1 ..`
* FEATURE: `dng help` now fully implemented
* FEATURE: more information added to `dng call` output
* BUGFIX: When running `dng call` with HTSLIB 1.2.1, the following error message was being
emitted: "FIXME: dirty header not synced".  This has been fixed.
* BUGFIX: `dng call` now outputs correct 1-based site locations.
* BUGFIX: a segfault was fixed in `dng dnm` and `dng phaser` caused by invalid commandline arguments
* CHANGE: HTSLIB 1.2+ is now required.
* Miscellaneous improvements to the build system

anderspitman committed:
Initial commit

anderspitman committed:
Basic tooling and structure setup

anderspitman committed:
Add test and cleanup a little

anderspitman committed:
Add test data file

anderspitman committed:
Basic parse button working

anderspitman committed:
Automatically load example vcf

anderspitman committed:
Add some graph parsing functionality

anderspitman committed:
Basic graph implmented

anderspitman committed:
Generations basically working

anderspitman committed:
Generalize generation calculation

Figuring out the generation of each node is no longer depending
on starting with a gen-1 node.

anderspitman committed:
Start process of using kinship2 for layout

anderspitman committed:
Basic round trip to server for R code

The user now uploads a pedigree file in the denovogear format.
The javascript server hands it off to R for pedigree creation
and layout, then hands a json object back to the client

anderspitman committed:
Snapshot. Basic nodes and links in place

anderspitman committed:
Basic child links in place

anderspitman committed:
Remove a bunch of commented code

anderspitman committed:
Allow uploaded ped to be used

Needed to delete the old chart if it existed

anderspitman committed:
Add pan and zoom behavior

anderspitman committed:
Basic hover display

Using bootstrap

anderspitman committed:
Switch back to running on localhost

anderspitman committed:
Plot in upper left corner

anderspitman committed:
Add pedigree graph class and ped parsing

anderspitman committed:
Use builder pattern for Person class

anderspitman committed:
DNG output plumbing

anderspitman committed:
Highlight selected node

anderspitman committed:
Link sample data to newick trees

anderspitman committed:
Add example_output.vcf

anderspitman committed:
snapshot

anderspitman committed:
Demo Draft 1

Very rough working demo. Shows mutation on pedigree. Currently
only tested on testdata/relationship_graph

anderspitman committed:
Add script to install R dependencies

anderspitman committed:
Handle empty newick column in ped file

anderspitman committed:
Basic working serverless version

Now depends on denovogear calling out to Rscript in order to
calculate the layout.

anderspitman committed:
Add gitignore and do some cleanup

anderspitman committed:
Improve development pipeline

Deployment and development pipelines now more closely match. Same
scripts are shared between the two.

Also added argument parsing.

anderspitman committed:
Remove build.js

Functionality rolled into build.py

anderspitman committed:
Restructure repository

Moved things around a lot. More structured hierarchy now.

anderspitman committed:
Port build.py to build.R

anderspitman committed:
Merge pull request #7 from anderspitman/bundle

Bundle

anderspitman committed:
Move everything into src/mutmap

anderspitman committed:
Merge remote-tracking branch 'dng-vis/move-to-subdirectory' into merge-dng-vis

dng-jenkins commented:
Can one of the admins verify this patch?

anderspitman committed:
Final housekeeping to merge repos

Manually pulled CMake files from the old visualization branch,
which will be defunct if this branch gets merged.

anderspitman committed:
Cleanup after repo merge

anderspitman commented:
One caveat to including the visualization stuff directly into denovogear is that we're creating a dev dependency on the javascript toolchain (node, etc). I'm assuming we want to add a flag to control whether to build the visualization functionality?

reedacartwright commented:
Jenkins, okay to test.

reedacartwright commented:
I'll do a code review.

reedacartwright commented on src/mutmap/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+find_package(Rscript)
+
```
Rscript should be an optional dependency.

anderspitman replied:
done

reedacartwright commented on src/mutmap/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+
+configure_file("${CMAKE_CURRENT_SOURCE_DIR}/dist/bundle.js"
+  "${CMAKE_CURRENT_BINARY_DIR}/bundle.js" @ONLY)
+
```
Node.js stuff needs to be removed.

anderspitman replied:
Node is required for javascript tooling. We would need to reinvent a lot of wheel to make up for the functionality.

reedacartwright replied:
Why do we need Javascript tooling? We just need a html template and associated .js files.

anderspitman replied:
The biggest reason currently is that I'm using browserify, which handles bundling all the different javascript source files together. But it also allows you to use the `require `syntax, which is very nice for using js in a modular way. Without it we would need to do all of the bundling ourselves. I can give it a shot and see how ugly it turns out, but I don't think it's worth it. Once you have node, there are tons of useful tools at your disposal.  It's just how things are done in js land. For example, I'd like to convert all my OOP code to typescript when I have a couple extra hours, to make it more maintainable in the future. Using typescript requires node.

reedacartwright replied:
You would be right if we were designing a web app. However, we are not designing a web app. We are designing an interactive report. The code does not need to be modular since we are unlikely to reuse it in other parts of DeNovoGear anytime soon. We also are already using a building/packaging system in CMake and should use that as much as we can to bundle source files together into the template system. Written correctly, it is as simple as "cat *.js > output.js".

We also likely don't need typescript support. Depending on node might make the developer's life easier, but adds an additional dependency to the use of DeNovoGear. And one that doesn't have anything to do with runtime benefits.

reedacartwright replied:
To see where I'm coming from, look at how I design web apps:

https://github.com/CartwrightLab/website/tree/master/apps

(You should have access to that directory.)

anderspitman replied:
Node is only a dependency for developers of denovogear, not a runtime dependency. True we are developing a static report, but that static report is a web app. It has all of the same functionality.

I will take a crack at setting everything up dependency free and see how it looks.

That link is giving me a 404.

reedacartwright replied:
Can you explain your workflow to me? I'm not sure how node is only required by the developer and not during compiling/building.

Since you don't have permission to look at that repo, you can try looking at the source of the WhiteZed app: https://cartwrig.ht/apps/whitezed/.

anderspitman replied:
Sorry I was assuming people compiling would be considered developers, but I guess most of our users compile from source. I agree a dependency on Node should be avoided, even though the visualization component will be optional.

I don't think the bundling will be too bad. I already have everything modularized I just need to convert the syntax to a vanilla js module pattern instead of the CommonJS semantics node uses.

anderspitman replied:
Alright node dependency should be gone now.

reedacartwright commented on src/mutmap/src/html/index.html:
```diff
@@ -0,0 +1,53 @@
+
+<script>
+<!--BUNDLE_JS_PLACEHOLDER-->
+</script>
```
No bundle.js

anderspitman replied:
same as above

anderspitman commented on .gitignore:
```diff
@@ -40,4 +40,3 @@ src/config.h
 # Ignore the build directory
 build/*
 !build/.gitkeep
-
```
not sure what the deal is with that '-'. Accidentally removed it. Is it necessary?

reedacartwright replied:
A blank line was removed somewhere in the process.

reedacartwright commented on src/mutmap/test/data/dng.vcf:
```diff
@@ -0,0 +1,125 @@
+##fileformat=VCFv4.2
```
Remove test data.  Testdata is in the testdata repository and is downloaded when `make test` runs.

reedacartwright commented on src/mutmap/tools/build.R:
```diff
@@ -0,0 +1,108 @@
+
+parseArguments <- function() {
+
+    library(argparser, quietly=TRUE)
```
argparser is a dependency.

anderspitman replied:
added to the dependency install script

reedacartwright commented on src/mutmap/tools/install_r_dependencies.R:
```diff
@@ -0,0 +1,2 @@
+install.packages("kinship2", repos='http://cran.us.r-project.org')
```
This should be part of documentation and an error message.

anderspitman replied:
What's the best place to add this in the docs?

reedacartwright replied:
https://github.com/denovogear/denovogear/wiki/Manual:-dng-mutmap

anderspitman committed:
Remove mutmap test data

anderspitman committed:
Remove node.js dependency

Was really only using it for bundling via browserify. Moved a
couple of the dependencies to html script tags and bundled our
code manually using CMake.

anderspitman committed:
Make Rscript and mutmap optional

anderspitman commented:
@reedacartwright I believe this is ready for another pass.

anderspitman committed:
Add simple server for autotesting

Since running the entire build.R script is pretty slow due to
loading the R dependencies, I added a function to the script that
runs a socket server. When an incoming connection is established,
it runs the test code. This provides a much faster turnaround
since the libraries only have to be loaded once when the script
starts.

anderspitman committed:
Set test server to infinite timeout

anderspitman committed:
Remove some commented prints

Also increased timeout for automated tester

anderspitman committed:
Remove bash shim for mutmap

Now calls the R script directly, instead of going through the
shell script

anderspitman committed:
Improve argument parsing for mutmap

anderspitman committed:
Make test server a standalone script

Can now run test_server.R and autotest.sh to have a complete
automated development testing environment.

anderspitman committed:
Clean up commented code in dng_vis.js

anderspitman committed:
Modularize mutmap code a bit

anderspitman committed:
Add linting and fix lint errors

anderspitman committed:
Add layout_and_template.R

anderspitman committed:
Do a little refactoring for clarity

anderspitman committed:
Snapshot

anderspitman committed:
Snapshot

anderspitman committed:
Add rough draft of sample tree visualizations

anderspitman committed:
Basic rendering of duplicates

Extra link to parents is gone and a nice dotted line is now
drawn between the duplicates, like kinship2 does.

anderspitman committed:
Nest sample trees in SVG

Less of a hack now. The sample trees are nested under the SVG
nodes. Still a bit of a hack with passing the data through but
it doesn't seem too bad.

anderspitman committed:
Add more display attributes

anderspitman committed:
Handle window resizing

anderspitman committed:
Process PED files with 5 columns

Interprets IDs as sample IDs

anderspitman committed:
Fix some click infos

anderspitman committed:
Way too many changes in this commit

* Added redux and immutable js and started handling some of
the state using those. It's a little easier to reason about now
and should get better moving forward.
* Added genome browser view for showing where in the contig
the mutations are. Still needs lots of work.
* Split some things into separate files.

anderspitman committed:
Move store into global module

anderspitman committed:
Snapshot

Broke some things but improved the organization a bit

anderspitman committed:
Implemented update without complete rerender

Broke resizing and automatic full height settings (forcing to a
set height for now) but overall working pretty well.

anderspitman committed:
Split stats view off into separate class/file

anderspitman committed:
Merge browser views

anderspitman committed:
Auto svg width and height

Also added full genome, but need to separate it to ensure
separate selections

anderspitman committed:
Remove jQuery dependency except for bootstrap

anderspitman committed:
Fix eslint errors

anderspitman committed:
Changed from Redux/Immutable to PubSub

Redux really seemed like overkill. We'll see how this scales
though...

anderspitman committed:
Change genome browser size

Also added border when highlighting mutations

anderspitman committed:
Made pedigree view take up available space

anderspitman committed:
Fixed stats bug

Stats view wasn't updating when mutations were clicked

anderspitman committed:
Start with pedigree centered

Also added fancy zoom in animation for no good reason

anderspitman committed:
Re-enable sample tree toggle

anderspitman committed:
Snapshot. Female circles mostly working

anderspitman committed:
Implemented ratio nodes

anderspitman committed:
Basic stroke-width based highlight

Good enough for now

anderspitman committed:
Greatly improved genome browser

anderspitman committed:
Implement zooming for genome browser

anderspitman committed:
Snapshot - broken

anderspitman committed:
Simple genome view for single contigs

anderspitman committed:
Highlight selected mutation

anderspitman committed:
Cycle through mutations and contigs

anderspitman committed:
Mutation drop down and direct click selection

anderspitman committed:
snapshot

anderspitman committed:
Fix svg size with latest d3

anderspitman committed:
Fix nodes not updated when locus changed

anderspitman committed:
Add table for GP length 10

anderspitman committed:
Merge branch 'master' of github.com:anderspitman/denovogear into merge-dng-vis

anderspitman committed:
Merge branch 'merge-dng-vis' of github.com:anderspitman/denovogear into merge-dng-vis

anderspitman committed:
Rename dng_vis to mutmap

anderspitman committed:
Make render relative to container

The mutation explorer view now renders into whatever container
it is provided, rather than grabbing ids from the HTML. As a
consequence, it now renders the HTML itself.

anderspitman committed:
Cleanup build a bit

* Renamed and moved view files into views directory
* Improved file change detection for autotest.sh

anderspitman committed:
Add tabs to interface

anderspitman committed:
Snap shot - fix height of pedigree

anderspitman committed:
Basic distribution pedigree implemented

anderspitman committed:
Snapshot

anderspitman committed:
Add bar charts for mutation counts

anderspitman committed:
Mutation distribution looking pretty good

anderspitman committed:
Make mutation distribution fill the screen

anderspitman committed:
Begin making custom VCF parser

anderspitman committed:
Custom VCF parser basically working

Currently faster than vcf.js, but still taking about 2.5 seconds
to parse the test file I'm using.

anderspitman committed:
Implement lazy evaluation for VCF record parsing

Huge speed improvement for our use case

anderspitman committed:
Cache additional properties

anderspitman committed:
Remove dependency on vcf.js

anderspitman committed:
Create factory methods

For PedigreeView and StatsView

anderspitman committed:
Fix GT NaN

anderspitman committed:
Build mutation location data and add tab

anderspitman committed:
Snapshot - mutation locations looking good

Have a good start on it

anderspitman committed:
Remove spurious length row

anderspitman committed:
Snapshot

anderspitman committed:
Snapshot

anderspitman committed:
Re-render mutation locations when data changes

anderspitman committed:
Display text of current chromosome

anderspitman committed:
View base class

ListSelectorView broken off into its own class and inherits from
View now.

anderspitman committed:
Don't switch tabs for debug

anderspitman committed:
Begin transition to Backbone

So far, so good. Converting ListSelectorView was pretty easy.
Backbone appears to be a good replacement for views and PubSub.

anderspitman committed:
Move trigger setup to intialize

Was being called on every update. Ended up being kind of a can of
worms, but it appears to be working now.

anderspitman committed:
Clean up a couple module definitions

anderspitman committed:
Convert more views to Backbone

Everything in mutation_locations.js

anderspitman committed:
Convert MutationDistributionView to Backbone

Also started on StatsView

anderspitman committed:
Convert StatsView to Backbone

anderspitman committed:
Start converting pedigree to Backbone

anderspitman committed:
Fixed extra nodes bug

Extra nodes were being added on redraw due to enter/update
selection being used instead of just enter

anderspitman committed:
Remove PubSub dependency from PedigreeView

anderspitman committed:
Convert ContigView to Backbone

Still need to remove PubSub dependency

anderspitman committed:
Remove PubSub dependency from ContigView

Also cleaned it up quite a bit, especially the interface
between MutationExplorerView and ContigView

anderspitman committed:
Finish removing PubSub dependency

Now completely switched over to using Backbone events

anderspitman committed:
Remove some unused and commented code

anderspitman committed:
Convert MutationExplorerView to Backbone

anderspitman committed:
Replace last vestiges of renderInto with el

anderspitman committed:
Add navbar

Still need to do the routing for it

anderspitman committed:
Switch to Backbone for routing

anderspitman committed:
Cache view renders

anderspitman committed:
Properly render views from sizes of containers

anderspitman committed:
Implement screen resizing

Currently broken for MutationLocationsView since it's render
function is adding new elements.

anderspitman committed:
Don't append extra MutationLocationsViews

Was happening on resize. Still need to properly implement resize
for the view.

anderspitman committed:
Remove optionsManager

anderspitman committed:
Remove mutation details from distribution view

anderspitman committed:
Fix some layout issues

anderspitman committed:
Clean up mutation locations view a bit

anderspitman committed:
Fix bug where node visuals weren't updating

anderspitman committed:
MutationLocationsView responsive to window resize

anderspitman committed:
Improve hack in MutationLocationListView

Works more closely with d3 model now

anderspitman committed:
Use model for passing data to SampleMutationsView

anderspitman committed:
Use Backbone model for MutationLocationsView

anderspitman committed:
Use svg:svg instead of svg:g for nesting

Simplifies getting and setting the width and height. g elements
have no width/height, they simply match their contents. So before
I was doing a hack by appending a rectangle of the size I wanted.
Instead, I'm now using nested svgs which seems to work fine.

anderspitman committed:
Start documentation for mutmap

anderspitman committed:
Snapshot

anderspitman committed:
Switch pedigree coordinates to ratios

Pedigree layout coordinates are now generated as ratios between
0 and 1. This makes it easier to compute node and link positions
later on in the rendering process.

anderspitman committed:
Center and zoom explorer pedigree

anderspitman committed:
Add data for variant distribution

anderspitman committed:
Implement ratio circle

anderspitman committed:
Implement ratio square

anderspitman committed:
Change ratio_square to more general rectangle

anderspitman committed:
Add margins to mutation distribution view

anderspitman committed:
Fix mutation location selector

anderspitman committed:
Change order of variant types

anderspitman commented:
@reedacartwright and I discussed the future of this visualization tool and decided it makes more sense to keep it as a separate repository for now, sort of like a plugin. That repository now lives [here](https://github.com/denovogear/dng-mutmap)
